docs/docs/design/grandpa-client.md

```diff
@@ -0,0 +1,272 @@
+
+Introduces `ChangeTree` which is a structure built to track pending changes across forks.  Introduces `AuthoritySet` which trackers the current authorities, as well associated any standard and/or forced authority set changes. 
+
+### client(consensus/grandpa): implement grandpaDb and environment types ([#3383](https://github.com/ChainSafe/gossamer/pull/3383))
+
+Implements the auxiliary data portion of the client GRANDPA integration.  Functionality like persisting the `voterSetState`, current authority set, best block justification to disk is implemented in this PR.
+
+### client(consensus/grandpa): implement grandpa justification logic ([#3454](https://github.com/ChainSafe/gossamer/pull/3454))
+
+Implements verification of GRANDPA justifications, and remaining auxiliary data persistance.
+
+### client(consensus/grandpa): implement finality proof logic ([#3589](https://github.com/ChainSafe/gossamer/pull/3589))
+
+Implements creation and verification of GRANDPA finality proofs.
+
+
+## What Remains?
+
+The following high level types, `BlockImport` trait implementation, and integration tests still need to be completed.
+
+### `Environment`
+
+[`Environment`] is a type that utilizes already introduced dependencies (`VoterSet`, `SharedAuthoritySet`, etc.) to provide functionality to the [`VoterWork`] type.  It is a self contained type that encapsulates all dependencies of the [`VoterWork`] type.
+
+### `VoterWork` functionality
+
+`sc_consensus_grandpa` main future is [`VoterWork`] type.  [`VoterWork`] uses `communication::NetworkBridge`, the grandpa `Voter` from the standalone GRANDPA package, and the [`Environment`] type as chief dependencies to process GRANDPA voter message and progress the standalone GRANDPA voter. Given the rust source utilizes async rust code, further exploration and prototyping needs to be done to implement it in a way that aligns with Go best practices with regards to concurrency.
+
+### `NetworkBridge`
+ [`NetworkBridge`] is the bridge between the underlying network service, gossiping consensus messages and GRANDPA. It contains an implementation of `Network` trait and an implementation of `Syncing` trait though these are just stored to instantiate [`GossipEngine`] which is a dependency.  There is also a [`GossipValidator`](https://github.com/paritytech/polkadot-sdk/blob/d2fd53645654d3b8e12cbf735b67b93078d70113/substrate/client/consensus/grandpa/src/communication/gossip.rs#L1486) type used as a dependency. 
+ 
+ I propose replicating [`NetworkBridge`] given it's a good abstraction and a defined dependency of `VoterWork`.   We will also need to replicate [`GossipEngine`] given it's a depency of [`NetworkBridge`]. 
+ 
+### `GossipEngine`
+
+[`GossipEngine`] is the main long running parallel process that provides generalized gossip functinonality in substrate.  It will also be required for BEEFY implementation. [`GossipEngine`] requires an implementation of the `Network` trait as well as the `Syncing` trait.
```

Comment:
Is there any evidence showing we need BEEFY to propose a block? I personally have not seen any, so if there is not I think we need not worry to much about BEEFY atm

Reply:
🤔 The `GossipEngine` sub-dependency is essentially what is used to drive protocol specific gossip functionality.  Nothing to do with proposing blocks.

Reply:
Yeah, I just wanted to make sure that this was needed specifically for grandpa and that we weren't just building it for beefy. I mention building blocks as it relates to our goals, not specifically GossipEngine, just wanting to make sure that it was necessary for our block-proposing goals.

Reply:
I'm still confused.  BEEFY has it's own sub protocol that relies on `GossipEngine`.  Whether or not it rolls up to our block proposing goals? I would say yes.  We need to gossip both BEEFY and GRANDPA subprotocols to participate as a validator.

Reply:
Resolved offline